chain/syncer_integration_test.go

```diff
@@ -27,400 +27,78 @@ import (
+	cborStore := hamt.CborIpldStore{Blocks: bserv.New(bs, offline.Exchange(bs))}
+	store := chain.NewStore(repo.ChainDatastore(), &cborStore, &state.TreeStateLoader{}, genesis.At(0).Cid())
+	require.NoError(t, store.PutTipSetAndState(ctx, &chain.TipSetAndState{genStateRoot, genesis}))
+	require.NoError(t, store.SetHead(ctx, genesis))
+
+	// Note: the chain builder is passed as the fetcher, from which blocks may be requested, but
+	// *not* as the store, to which the syncer must ensure to put blocks.
+	eval := &chain.FakeStateEvaluator{}
+	syncer := chain.NewSyncer(eval, store, builder, chain.Syncing)
+
+	base := builder.AppendManyOn(3, genesis)
+	left := builder.AppendManyOn(4, base)
+	right := builder.AppendManyOn(3, base)
+
+	// Sync the two branches, which stores all blocks in the underlying stores.
+	assert.NoError(t, syncer.HandleNewTipset(ctx, left.Key(), ""))
+	assert.NoError(t, syncer.HandleNewTipset(ctx, right.Key(), ""))
+	verifyHead(t, store, left)
+
+	// The syncer/store assume that the fetcher populates the underlying block store such that
+	// tipsets can be reconstructed. The chain builder used for testing doesn't do that, so do
+	// it manually here.
+	for _, tip := range []types.TipSet{left, right} {
+		for itr := chain.IterAncestors(ctx, builder, tip); !itr.Complete(); require.NoError(t, itr.Next()) {
+			for _, block := range itr.Value().ToSlice() {
+				_, err := cborStore.Put(ctx, block)
+				require.NoError(t, err)
+			}
+		}
 	}
 
-	forklink1blk1 := th.RequireMkFakeChild(t, fakeChildParams)
-
-	fakeChildParams.Nonce = uint64(1)
-	forklink1blk2 := th.RequireMkFakeChild(t, fakeChildParams)
-
-	fakeChildParams.Nonce = uint64(2)
-	forklink1blk3 := th.RequireMkFakeChild(t, fakeChildParams)
-	forklink1 := th.RequireNewTipSet(t, forklink1blk1, forklink1blk2, forklink1blk3)
-
-	fakeChildParams.Parent = forklink1
-	fakeChildParams.Nonce = uint64(0)
-	forklink2blk1 := th.RequireMkFakeChild(t, fakeChildParams)
-
-	fakeChildParams.Nonce = uint64(1)
-	forklink2blk2 := th.RequireMkFakeChild(t, fakeChildParams)
+	// Load a new chain store on the underlying data. It will only compute state for the
+	// left (heavy) branch. It has a fetcher that can't provide blocks.
+	newStore := chain.NewStore(repo.ChainDatastore(), &cborStore, &state.TreeStateLoader{}, genesis.At(0).Cid())
+	require.NoError(t, newStore.Load(ctx))
+	offlineSyncer := chain.NewSyncer(eval, newStore, chain.NewBuilder(t, address.Undef), chain.Syncing)
 
-	fakeChildParams.Nonce = uint64(2)
-	forklink2blk3 := th.RequireMkFakeChild(t, fakeChildParams)
-	forklink2 := th.RequireNewTipSet(t, forklink2blk1, forklink2blk2, forklink2blk3)
+	assert.True(t, newStore.HasTipSetAndState(ctx, left.Key().String()))
+	assert.False(t, newStore.HasTipSetAndState(ctx, right.Key().String()))
 
-	fakeChildParams.Nonce = uint64(0)
-	fakeChildParams.Parent = forklink2
-	forklink3blk1 := th.RequireMkFakeChild(t, fakeChildParams)
+	newRight := builder.AppendManyOn(2, right)
```

Comment:
Blocking: these new blocks need to be available within the offlineSyncer's fetcher.  Right now they are placed in the original fetcher so the offlineSyncer will just fail because it couldn't fetch the two new blocks.  This test should instead check that the `HandleNewTipset` call fails because the *`right`* chain is not fetchable or considered already "stored" in the chain store even though all the blocks are kept in the `Store`'s internal repo.

In the previous version this intent is captured by putting `link1` and `link2`'s blocks to the cborStore while putting `link3`'s blocks to the `loadSyncer`'s `blockSource`.

Reply:
Ok, I think I fixed it, and have written an essay attempting to communicate this. 

I think this is a bad test. In particular, I think it would fail if the code and syncer/store interactions were correct, but just different to how they are now - i.e. if the syncer could read individual blocks, but didn't make invalid inferences about the state it has thus computed.

The interactions of the syncer, fetcher, and store are weird - I wouldn't yet call it "detangled", even after we clean up these tests. I will write more in an issue.

Reply:
@anorth as I mentioned [here](https://github.com/filecoin-project/go-filecoin/pull/3094#discussion_r305959284) we should consider ourselves free to remove this test.